### PR TITLE
CNAO integration

### LINKED ISSUE #125: Use go template with cluster-network-addons-operator format for manifests
So we can work this https://github.com/kubevirt/cluster-network-addons-operator/issues/178

### PR DESCRIPTION
**Pull request by qinqon**
Signed-off-by: Quique Llorente <ellorent@redhat.com>

**What this PR does / why we need it**:
This PR use go template formant from CNAO so it's easier to take directly the manifests from here and use it there instead of manual copying it.

It also make cluster-sync more robust by implementing consistently and eventually with timeout/interval mechanism.

Also update kubevirtci to remove transient error about hostname not set by dhclient yet.

Closes #125
**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: hack/render-manifests.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	}
+	inputDir := os.Args[4]
+	outputDir := os.Args[5]
+
+	tmpl, err := template.ParseGlob(inputDir + "/*.yaml")
+	if err != nil {
+		exitWithError(err, "failed parsing manifests at %s", inputDir)
+	}
+	for _, t := range tmpl.Templates() {
+
```
Comment (phoracek) [inline hack/render-manifests.go]: Drop

**Comment by qinqon** [comment]
Let's merge first https://github.com/nmstate/kubernetes-nmstate/pull/442

File: Makefile [hunk 2]
```diff
@@ -173,11 +167,11 @@ tools-vendoring:
 	test/e2e \
 	cluster-up \
 	cluster-down \
-	cluster-sync-resources \
```
Comment (phoracek) [inline Makefile]: Sorry about the leftover. The following line can be dropped too

File: cluster/sync.sh [hunk 3]
```diff
@@ -2,80 +2,106 @@
...
 }
 
 function getNumberAvailable {
-    numberAvailable=$(./cluster/kubectl.sh get daemonset -n nmstate $1 -o=jsonpath='{.status.numberAvailable}')
+    numberAvailable=$($kubectl get daemonset -n nmstate $1 -o=jsonpath='{.status.numberAvailable}')
     echo ${numberAvailable:-0}
 }
 
+function eventually {
+    timeout=15
```
Comment (phoracek) [inline cluster/sync.sh]: These changes are unrelated to the change in generating of manifests, right? Please open a dedicated PR for that.

File: cluster/sync.sh [hunk 4]
```diff
@@ -2,80 +2,106 @@
...
+            registry=localhost:$(./cluster/cli.sh ports --container-name=cluster registry | tr -d '\r')
     else
-        return 1
+        registry=localhost:$(./cluster/cli.sh ports registry | tr -d '\r')
     fi
-}
 
-# Cleanup previous deployment, if there is any
-make cluster-clean
-
```
Comment (phoracek) [inline cluster/sync.sh]: I like the refactoring, but it is really difficult to review this PR since it contains multiple unrelated changes. Please move this refactoring to a separate PR or at least a dedicated commit. Let's do it properly

File: cluster/sync.sh (hunk 3 above)
Comment (qinqon) [reply cluster/sync.sh]: We need them to make CI happy, and they change same script, we don't need to backport this since we don't have webhooks at release-0.15

File: cluster/sync.sh (hunk 4 above)
Comment (qinqon) [reply cluster/sync.sh]: ooook

**Comment by qinqon** [comment]
/hold

Let's split sync.sh refactor so we can merge before it or even backport it.

**Comment by qinqon** [comment]
/retest

We may need the timeout to check transient hostname

File: Makefile [hunk 5]
```diff
@@ -96,7 +95,10 @@ gen-openapi: $(OPENAPI_GEN)
 gen-crds: $(OPERATOR_SDK)
 	$(OPERATOR_SDK) generate crds
 
-handler: gen-openapi gen-k8s gen-crds $(OPERATOR_SDK)
+manifests:
+	$(GO) run hack/render-manifests.go nmstate $(HANDLER_IMAGE) Always deploy/ $(manifests)
```
Comment (phoracek) [inline Makefile]: Please set all the attributes via a changeable env variable. I don't have a good usecase for it now, but it may save us some PR iterations in the future if we need to generate manifests for different namespace for instance.

File: cluster/sync.sh [hunk 6]
```diff
@@ -60,9 +59,22 @@ function deploy() {
         registry=localhost:$(./cluster/cli.sh ports registry | tr -d '\r')
     fi
 
+    manifests_dir=build/_output/manifests
```
Comment (phoracek) [inline cluster/sync.sh]: Since this is defined on multiple places, please define it on Makefile and pass it as env variable to these scripts.

File: hack/render-manifests.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+		ImagePullPolicy     string
+	}
+
+	inventory := Inventory{
+		Namespace:           os.Args[1],
+		NMStateHandlerImage: os.Args[2],
+		ImagePullPolicy:     os.Args[3],
+	}
+	inputDir := os.Args[4]
+	outputDir := os.Args[5]
```
Comment (phoracek) [inline hack/render-manifests.go]: Please count passed parameters first so we can fail with a user-friendly error if they don't provide enough arguments.

File: hack/render-manifests.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+	if err != nil {
+		exitWithError(err, "failed cleaning up output dir %s", outputDir)
+	}
+
+	tmpl, err := template.ParseGlob(path.Join(inputDir, "*.yaml"))
+	if err != nil {
+		exitWithError(err, "failed parsing top dir manifests at %s", inputDir)
+	}
+
+	tmpl, err = tmpl.ParseGlob(path.Join(inputDir, "*/*.yaml"))
```
Comment (phoracek) [inline hack/render-manifests.go]: please use better variables  for tmpl, I don't understand why you override it now.

File: hack/render-manifests.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+	if err != nil {
+		exitWithError(err, "failed parsing sub dir manifests at %s", inputDir)
+	}
+
+	for _, t := range tmpl.Templates() {
+		manifestDir := outputDir
```
Comment (phoracek) [inline hack/render-manifests.go]: Why do we need this?

File: hack/render-manifests.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+		// We don't have the full path from ParseGlon so we just guess here
+		// this is not going to change too much
+		if strings.Contains(t.Name(), "scc.yaml") {
+			manifestDir = path.Join(manifestDir, "openshift")
+		} else if strings.Contains(t.Name(), "_crd.yaml") || strings.Contains(t.Name(), "_cr.yaml") {
+			manifestDir = path.Join(manifestDir, "crds")
```
Comment (phoracek) [inline hack/render-manifests.go]: Do we need subdirectories at all? Can't we just keep it in a single one?

Comment (qinqon) [reply hack/render-manifests.go]: Just one to make it similar ot template dir so we don't have to change much stuff at release scripts.

File: Makefile (hunk 5 above)
Comment (qinqon) [reply Makefile]: Ok, seting nmstate as env var, pullPolicy not so, it's ok ?

File: hack/render-manifests.go (hunk 8 above)
Comment (phoracek) [reply hack/render-manifests.go]: @qinqon

File: hack/render-manifests.go (hunk 9 above)
Comment (qinqon) [reply hack/render-manifests.go]: I have refactor it to put everything at same directory.

File: hack/render-manifests.go (hunk 10 above)
Comment (qinqon) [reply hack/render-manifests.go]: Have refactor and now is everything at the same dir.

File: hack/render-manifests.go (hunk 8 above)
Comment (phoracek) [reply hack/render-manifests.go]: I see now. Thanks
